### ISSUE TITLE
High DPI scaling factor/Wrong monitor dimensions

### ISSUE THREAD

**Issue by NellyWhadsDev** [jwilm/alacritty#1378]
Which operating system does the issue occur on?
Ubuntu 18.04

If on linux, are you using X11 or Wayland?
(Whatever it defaults to in Ubuntu 18.04 - the output of `loginctl` was a little confusing: it showed gdm was running in Wayland, and I was logged in in X11)
![screenshot from 2018-06-15 15-57-50](https://user-images.githubusercontent.com/8030181/41487111-4d7bafe0-70b5-11e8-87b5-d4e2178c30d0.png)

X11

When I launch alacritty, it seems to set a device_pixel_ration of 1.6666...
This means my fonts show up huge on the terminal window as soon as I start it. I only have one display attached to my laptop (the laptop's inbuilt display).

I know there are other issues posted here surrounding similar DPI scaling issues, but they all seemed to be around dynamic scaling based on which monitor Alacritty is launched from. Once again, I only have one; 1920*1080.

Firstly, is there a hot-fix which allows me to override the device_pixel_ratio somewhere in the alacritty.yml (currently all values are the defaults)?
Secondly, how can I find out where along the chain this is going wrong? The OS seems to see the 1920*1080 resolution just fine.

![screenshot from 2018-06-15 15-55-29](https://user-images.githubusercontent.com/8030181/41487123-5cc0a05a-70b5-11e8-969e-6bda7e6c9f65.png)

**Comment by NickeZ** [comment]
I had the same experience. I had to roughly set my font size to half the previous font size. So I have it at around 6-7 now.

**Comment by NellyWhadsDev** [comment]
Setting text size lower isn't necessarily the right solution; I can scale the text size from the config and using keybaord shortcuts, but it still roots from there being a faulty scaling factor coming from somewhere along the chain (`winit`?) in the latest LTS (18.04).

**Comment by nixpulvis** [comment]
I'm not 100% sure of anything related to DPI atm, but I'm pretty sure we're all waiting on tomaka/winit#548 before things will work as expected.

**Comment by NellyWhadsDev** [comment]
@nixpulvis It's been merged to master at least, just today; how long do you think we can expect to wait for everything to patched up all the way up to alacritty?

**Comment by nixpulvis** [comment]
I really have no idea, but the powers at play seem motivated :smile:

**Comment by nixpulvis** [comment]
Sorry won't happen again. I generally try to avoid gendered names with people online who I don't know, but my predisposition sometimes leaks through.

**Comment by chrisduerr** [comment]
`
now all my Alacritty instances across DPIs and screen sizes are all the same (physically), but they aren't in line with the way other things are scaled
`

I believe this was exactly the point of the change. Winit reports the accurate DPI rather than the screen's DPI which will very likely be wrong. If anyone doesn't want proper DPI but no DPI scaling, they can disable DPI scaling and set it through the font size. If people want a specific DPI for all instances they can use `WINIT_HIDPI_FACTOR=... alacritty`.

**Comment by chrisduerr** [comment]
`Winit reports the accurate DPI rather than the screen's DPI which will very likely be wrong.`

That's exactly this part. The DPI set it `xrandr` is set for each screen. This is likely going to be wrong on a per-monitor basis unless there's only one monitor. Which is why winit just ignores it and uses the proper DPI instead.

**Comment by nixpulvis** [comment]
Ok, thanks for the confirmation @chrisduerr. I'm assuming other platforms than X11 implement per screen (or monitor I guess) DPI correctly, does this mean those platforms won't receive proper scaling? I understand that there isn't a way to ask X11 for a DPI per monitor, but it still seems possible to scale with `xrandr --dpi XXX` if we just use that as a base factor, since if my understanding is correct, we **are** able to get correct per monitor DPIs by calculating them ourselves. 

Also, [this comment](https://github.com/tomaka/winit/issues/169#issuecomment-385504214), leads me to believe that the `Xft.dpi` variable will still work with Alacritty once everything is merged, is this correct?

**Comment by chrisduerr** [comment]
Basing the DPI we calculate ourselves off the DPI returned from the xrandr screen would be both wrong for the screen DPI way to handle DPI and for the self-calculated way to compute DPI. So instead of going with one or the other, we'd be inaccurate in both ways. The way it works right now is perfectly functional I'd say, it works the way DPI is supposed to work and the only issues is that other applications don't properly scale with a monitor's DPI. If people want to set custom DPIs, they can do so using `WINIT_HIDPI_FACTOR` or by disabling the dpi and setting the font. So I don't really see an issue.

Supporting `Xft.dpi` makes sense, but I'm not sure if that's part of winit yet.

Regarding other platforms there are just different issues with DPI, I believe with Wayland there isn't even an application DPI but the display server just renders everything at the same DPI and then scales it down to the monitor size. Which seems even worse than what X11 does but probably easier for application developers. I think it's also possible to set this up on X11 though, so that's another option. I don't know about any platform other than X11/Wayland though.

**Comment by NellyWhadsDev** [comment]
Current working solution (for anyone who runs into the same issue for any reason):

Added an `export` to me `.profile` which sets the `WINIT_HIDPI_FACTOR` to `1.0`. This now works (previously wasn't working) since #1373 was merged into master; relies on winit 0.15.1 insteawd of the previous 0.15.0.

Cheers folks; hope a true solution emerges soon!

**Comment by chrisduerr** [comment]
@NellyWhadsDev A `true solution` was already merged. If you always want your DPI to be fixed at `1.0`, you can use the following config setting:
https://github.com/jwilm/alacritty/blob/master/alacritty.yml#L96-L100

Hope that helps! For everything other than 1.0 DPI, the recommended way is still to use `WINIT_HIDPI_FACTOR`. However since all DPI does is changing the font size, if you want anything other than 1.0 DPI, you can just use 1.0 DPI and then tweak the font size.

**Comment by francesca64** [comment]
> However since all DPI does is changing the font size

That's not true. As of winit 0.16.0, window sizes are in logical pixels.

**Comment by chrisduerr** [comment]
Oh okay, that's news to me. Then using the `WINIT_HIDPI_FACTOR` would probably be the only way to go about this. Is there still any way to set the size in "real" pixels @francesca64? Because I'm pretty sure alacritty sets the size once the glyph cache is filled and the font metrics are present, so we just set the size to be `glyph_width * cols_requested_in_config + 2 * padding_x`, if winit adjusts window sizes now we'd probably have to change this.

**Comment by francesca64** [comment]
winit conserves logical sizes across DPI changes, so attempting to forcefully set physical size (by just multiplying the size you pass accordingly) would result in you needing to continuously override said size in the handler of `HiDpiFactorChanged`. The math for doing that is straightforward if you really want that, but I'd prefer if you don't want that (it's an anti-pattern).

**Comment by chrisduerr** [comment]
Hmm it should be possible to calculate the DPI-based size using something like `(glyph_width * cols_requested_in_config + 2 * padding_x) * dpr` I guess. I'll have to play around with this when winit 0.16 is released.

**Comment by chrisduerr** [comment]
Closing since it looks like the DPI calculated was correct, for overriding the DPI the environment variable `WINIT_HIDPI_FACTOR` can be used.

**Comment by nixpulvis** [comment]
In case anyone else stumbles back upon this issue, use the following (see https://github.com/alacritty/alacritty/issues/1501#issuecomment-521519997):

```toml
[env]
WINIT_X11_SCALE_FACTOR = "1"
```
